### PR TITLE
EAR-1004 Move Question Into And Out Of A Folder

### PR DESCRIPTION
**Pull request by jchapman68**
### What is the context of this PR?

Allows you to move questions into and out of folders. 

This slightly extends the current splicing method we use to add a question into a new position. This method doesn't really work with a non-flat structure that we have with folders now.

It might be reworked in the future depending on what Matt B and Joe discuss.

This PR includes:
- moving down into a folder
- moving up into a folder
- moving down out of a folder
- moving up out of a folder
- jumping below a folder when clicking on a folder and going down
- jumping above a folder when clicking on a folder and going up

**Moving down refers to starting above the position in which you intend to move too**
**Moving up refers to starting below the position in which you intend to move too**

### How to review
- Create a questionnaire with a question, folder with 3 questions and questions after the folder.
- Try to replicate the actions outlined above.

### CODE REVIEW

**Review by jasehumphr — APPROVED** [review]
Works wonderfully

**Review by martyncolmer — CHANGES_REQUESTED** [review]
I have a questionnaire with 
Q1
Q2
F1

Trying to move Q1 to after Q2 then back to start and then after again. It jumps to the end.

**Comment by jasehumphr** [comment]
> I have a questionnaire with
> Q1
> Q2
> F1
> 
> Trying to move Q1 to after Q2 moves it to after F1.

I don't get that

File: eq-author/src/components/PositionModal/index.js [hunk 1]
```diff
@@ -32,64 +32,123 @@ const Trigger = styled.button.attrs({ type: "button" })`
...
+      item: orderedOptions[previousPosition],
+    });
   };
 
-  const handleOpen = () => setIsOpen(true);
-
-  const handleChange = ({ value }) => setSelectedPosition(parseInt(value, 10));
+  const handleChange = ({ value }) =>
+    setOption(({ item: previousItem }) => {
+      let count =
```
Comment (MattElcock) [inline eq-author/src/components/PositionModal/index.js:66]:
```suggestion
      const count =
```

I think? I can't see it being modified anywhere else, but I might be wrong.

File: eq-author/src/components/PositionModal/index.js [hunk 2]
```diff
@@ -32,64 +32,123 @@ const Trigger = styled.button.attrs({ type: "button" })`
...
+
+    const positionCalculation = isSame
+      ? sameMove
+      : differentSection
+      ? folderMove
+      : fromFolderToFolder
+      ? folderMove
+      : inOrOut
+      ? folderMove
+      : itemPosition;
```
Comment (MattElcock) [inline eq-author/src/components/PositionModal/index.js:109]: Could this be changed to not use short hand, please? It's difficult to understand and I think we would benefit from having it use the tranditional if statements instead for if/when we read back over this.

**Review by MattElcock — CHANGES_REQUESTED** [review]

**Review by martyncolmer — DISMISSED** [review]
Adding a question to the root puts the new question inside the un-abled folder. The question should go in a folder on it's own.

**Review by MattElcock — COMMENTED** [review]
Cheers for making those changed. Happy to tick when that other bug is sorted.

**Comment by tbaggaley** [comment]
54om4n

File: eq-author/src/components/PositionModal/index.js [hunk 3]
```diff
@@ -32,64 +32,126 @@ const Trigger = styled.button.attrs({ type: "button" })`
...
+      positionCalculation = sameMove > 0 ? sameMove : 0;
+    } else if (differentSection || fromFolderToFolder || inOrOut) {
+      positionCalculation = folderMove;
+    } else {
+      const selectedItemPosition = orderedOptions.findIndex(
+        ({ id }) => id === selected.id
+      );
+      positionCalculation = selectedItemPosition;
+    }
+
```
Comment (tbaggaley) [inline eq-author/src/components/PositionModal/index.js:113]: Could you add a couple wee comments to this bit to explain what the logic behind all this arithmetic is? Could save Shane a headache when it comes to adding in moving folders ;-)

Comment (tjamesmac) [reply eq-author/src/components/PositionModal/index.js]: Will do, I mean it looks self explanatory to me... ;)

Comment (tjamesmac) [reply eq-author/src/components/PositionModal/index.js]: this i solved this

**Review by farres1 — DISMISSED** [review]

File: eq-author/src/App/page/Design/MovePageModal/index.test.js [hunk 4]
```diff
@@ -89,3 +99,72 @@ describe("MovePageModal: questionnaire not loaded", () => {
...
+    expect(output.filter(({ id }) => id.startsWith("folder-1-page"))).toEqual(
+      expect.arrayContaining([
+        expect.objectContaining({
+          parentEnabled: true,
+          parentId: output[0].id,
+          position: expect.any(Number),
+        }),
+      ])
+    );
+  });
```
Comment (MattElcock) [inline eq-author/src/App/page/Design/MovePageModal/index.test.js:164]:
```suggestion
  it("should include parent details on enabled folders", () => {
  
    const enabledFolders = output.filter(({ id }) => id.startsWith("folder-1-page"));
    expect(enabledFolders).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          parentEnabled: true,
          parentId: output[0].id,
          position: expect.any(Number),
        }),
      ])
    );
  });
```

The assertion of this test is slightly longer than the last, so you could toss the enabled folders into a variable to make it a little nicer on the eye.

File: eq-author/src/components/PositionModal/index.js [hunk 5]
```diff
@@ -32,64 +32,117 @@ const Trigger = styled.button.attrs({ type: "button" })`
...
+  const positionButtonId = uniqueId("PositionModal");
   const [isOpen, setIsOpen] = useState(false);
-  const [selectedPosition, setSelectedPosition] = useState(previousPosition);
-  const [positionButtonId] = useState(uniqueId("PositionModal"));
+  const previousIndex = options.findIndex(({ id }) => id === selected.id);
+  const previousPosition = useRef(previousIndex > -1 ? previousIndex : 0);
+  const [{ position, item }, setOption] = useState({
+    position: previousPosition.current,
+    item: options[previousPosition.current],
+  });
```
Comment (MattElcock) [inline eq-author/src/components/PositionModal/index.js:47]: I'm wondering if the variable names could be improved here. For instance:

`previousPosition` suggests it contains an integer, but actually it's the previous item in the list (I think). You could consider changing it to something like `previousItemRef`.

`previousIndex` becomes `previousItemIndex`.

```suggestion
  const previousItemIndex = options.findIndex(({ id }) => id === selected.id);
  const previousItemRef = useRef(previousItemIndex > -1 ? previousItemIndex : 0);
  const [{ position, item }, setOption] = useState({
    position: previousItemRef.current,
    item: options[previousItemRef.current],
  });
```

File: eq-author/src/components/PositionModal/index.js [hunk 6]
```diff
@@ -32,64 +32,117 @@ const Trigger = styled.button.attrs({ type: "button" })`
...
-  const handleOpen = () => setIsOpen(true);
-
-  const handleChange = ({ value }) => setSelectedPosition(parseInt(value, 10));
+  const handleChange = ({ value }) => {
+    const count =
+      orderedOptions[value].__typename === "Folder" && value - position >= 0
+        ? orderedOptions.filter(
+            ({ parentId }) => parentId === orderedOptions[value].id
+          ).length
+        : 0;
```
Comment (MattElcock) [inline eq-author/src/components/PositionModal/index.js:74]: You could split this up a bit to improve readability.

```suggestion
    const filteredOptions = orderedOptions.filter(
            ({ parentId }) => parentId === orderedOptions[value].id
          );
          
    const count =
      orderedOptions[value].__typename === "Folder" && value - position >= 0
        ? filteredOptions.length : 0;
```

**Review by MattElcock — APPROVED** [review]

File: eq-author/src/components/PositionModal/index.js (hunk 5 above)
Comment (tjamesmac) [reply eq-author/src/components/PositionModal/index.js:47]: previousPosition is an integer, `findIndex` returns a number

**Review by farres1 — APPROVED** [review]

**Review by jasehumphr — APPROVED** [review]

**Review by MattElcock — APPROVED** [review]

**Pull request merged**